### ISSUE TITLE
Crafty.unbind() should support a global unbind

### ISSUE THREAD

**Issue by cawoodm** [craftyjs/Crafty#417]
The core #Crafty.unbind always requires a callback. However it is sometimes necessary to unbind any and all callbacks from an event. I recommend the following change (line 1203) from:

``` javascript
if (hdl[h][i] === callback) {
```

to

``` javascript
if (hdl[h][i] === callback || typeof callback === 'undefined') {
```

**Comment by mpetrovic** [comment]
how are those lines different?

**Comment by GNi33** [comment]
@mpetrovic It's weird, they are the same in the sent notification-email, view the issue directly in the browser.

**Comment by mpetrovic** [comment]
Where are you seeing that line in here?
https://github.com/craftyjs/Crafty/blob/master/src/core.js#L528

**Comment by GNi33** [comment]
Why are there two .unbind - functions anyhow? How are they different?
https://github.com/craftyjs/Crafty/blob/master/src/core.js#L1194

**Comment by mpetrovic** [comment]
one of them is used on entities, the other on the global namespace.

**Comment by sbyrnes321** [comment]
Reminder -- this is the `handlers` object
https://github.com/craftyjs/Crafty/blob/91cca678b522d01ed8784920d9412a0915a29784/src/core.js#L1173

Let's take the example

```
handlers ===
  { Move: {5:[fnA], 6:[fnB, fnC], global:[fnD]},
  Change: {6:[fnE]}
  }
```

Now someone runs `Crafty.unbind('Move')`. What do we want to happen? GNi33 wants the result to be:

```
handlers ===
  { Change: {6:[fnE]}
  }
```

but I can imagine that in other situations someone would want the result to be

```
handlers ===
  { Move: {5:[fnA], 6:[fnB, fnC]},
  Change: {6:[fnE]}
  }
```

It seems to me that we need two different functions for these two different behaviors, maybe `Crafty.unbind_all(event)` and `Crafty.unbind_global(event)` or whatever.

By the way, not quite on topic, but the behavior of `Crafty.unbind('Move',0)` is not only strange but browser dependent ... it will delete the binding to either `fnA`, `fnB`, or `fnD`, depending on the order of iteration in the `for ... in` loop! Without the `return true` part, it would delete all three of them. Why would anyone want to do either of those things?? And by the way, wouldn't splicing things out of the arrays make more sense then deleting them? It could cause performance problems if you get an `handlers['Move']` array of `length` 10000 where the first 9999 elements are `undefined` ... and then you loop through it with a `for` loop ...

**Comment by mpetrovic** [comment]
This is getting confusing. Can someone put together a list of all the
possible unbinding scenarios and which ones we currently support?

As for the performance issue, I don't believe it's an issue because we're
not using actual arrays, but objects. We won't be iterating over 9999
undefined values.

**Comment by cawoodm** [comment]
Sorry for the confusion, I had initially pasted the same code twice.

What I am saying is that the _global_ unbind should offer the _additional_ feature of unbinding all listeners when no callback is passed. Otherwise it is impossible to remove bindings in cases where the listener was anonymous or is no longer references. This has no major side-effects since passing no callback today has no effect.

Old code:

``` javascript
unbind: function(event, callback) {
  var hdl = handlers[event],
    h, i, l;

  //loop over every object bound
  for (h in hdl) {
    if (!hdl.hasOwnProperty(h)) continue;

    //if passed the ID
    if (typeof callback === "number") {
      delete hdl[h][callback];
      return true;
    }

    //loop over every handler within object
    for (i = 0, l = hdl[h].length; i < l; i++) {
      if (hdl[h][i] === callback) {
        delete hdl[h][i];
        return true;
      }
    }
  }

  return false;
}
```

New code:

``` javascript
unbind: function(event, callback) {
  var hdl = handlers[event],
    h, i, l, res = false;                                               // Diff 1

  //loop over every object bound
  for (h in hdl) {
    if (!hdl.hasOwnProperty(h)) continue;

    //if passed the ID
    if (typeof callback === "number") {
      delete hdl[h][callback];
      return true;
    }

    //loop over every handler within object
    for (i = 0, l = hdl[h].length; i < l; i++) {
      if (hdl[h][i] === callback || typeof callback === 'undefined') {           // Diff 2
        delete hdl[h][i];
        res = true;                                               // Diff 3
      }
    }
  }

  return res;                                               // Diff 4
}
```

Diff:
Line 3: added `, res = false`
Line 17: added `|| typeof callback === 'undefined'`
Line 19: Replaced `return true;` with `res = true;`
Line 24: Replaced `return false;` with `return res;`

**Comment by cawoodm** [comment]
Also, can anyone tell me how to override the global `unbind` function? Everything I have tried has failed because `handlers` is private. It would be nice to know how to override or extend `Crafty` with our own methods which can access `handlers`.

For example this doesn't work:

``` javascript
Crafty.extend({
    unbind: function(event, callback) {
        var hdl = handlers[event]; // <-- Reference error here
    }
});
```

**Comment by mpetrovic** [comment]
you'll have to edit the source, due to the way scoping works in javascript.

You should submit a pull request to github so a fix can be included in the
next release.

**Comment by cawoodm** [comment]
@GNi33 one is for global events (`Crafty.bind()`) and the other for entities `myEntity.bind()`. When to use which is a good question and probably the resolution to many of my current woes. One would hope that when an entity is destroyed it's listeners are unbound. I can tell you now that the global `bind()` remains active even if the entity which bound them is destroyed (because of JavaScript closure) and can cause undesirable effects.

Agree with all of what @sbyrnes321 said. I hope my removal of `return true;` doesn't hurt the `unbind('MyEvent', ID` case because I need a function which will unbind _all_ listeners to an event.

@mpetrovic here is where I mean: https://github.com/craftyjs/Crafty/blob/master/src/core.js#L1179

**Comment by sbyrnes321** [comment]
@cawoodm -- If you want to fuss with the `handlers` object in a way that the available interface functions don't support, I suggest adding the line `Crafty.handlers = handlers;` after `handlers = {};` in the Crafty source. After that change, you can do whatever you want anywhere in your code by modifying `Crafty.handlers`. It's also helpful for debugging ... `console.log(Crafty.handlers);`.

**Comment by sbyrnes321** [comment]
Here is my proposal for what this function ought to do:
- `Crafty.unbind(eventName, 'all');` -- unbind all listeners from eventName, both globally-attached and entity-attached
- `Crafty.unbind(eventName, 'global');` -- unbind all the globally-attached event listeners from eventName
- `Crafty.unbind(eventName, functionObject);` -- search through both the entity-attached and globally-attached event listeners for all references to this particular function object, and erase them.
- `Crafty.unbind(eventName, number);` -- does nothing. The current behavior is not only useless but dangerous (being cross-browser-incompatible).
- `Crafty.unbind(eventName);` -- does nothing, just like the current situation

The function should return true if anything was erased, or false if nothing was erased.

The deletions within arrays should be splice-outs. Despite @mpetrovic 's comment above, the sub-sub-components of handlers, e.g. `handlers['Move']['global']`, really are arrays. They inherit from `Array` and are treated as arrays by iterating with `for` not `for ... in` both in this function and elsewhere in the code base. More importantly, `some_entity.unbind(...)` works by splicing not deleting, so we're just making `Crafty.unbind(...)` consistent.

**Comment by cawoodm** [comment]
Thanks for that tip @sbyrnes321!

Regarding your suggestion - I personally don't find passing a `functionObject` nor a `number` useful at all. In most cases you just want the event unbound and don't want to be bothered with keeping a reference to functions in order to unbind them. I think `Crafty.unbind(eventName)` is sufficient for this and should do what your `'all'` parameter does. IMHO. But I'm just happy if we get this feature regardless of how.

**Comment by mpetrovic** [comment]
I think we're letting Crafty.unbind get a little too powerful here.

Crafty.unbind should only handle global handlers. It should not be touching
handlers on individual entities. This would keep it consistent with
Crafty.bind, and make it the reverse of Crafty.bind().
With this restriction in mind, we can limit it to 2 scenarios:

Crafty.unbind(eventName)
Crafty.unbind(eventName, function)

Unbinding all instances of an event is something we probably shouldn't be
doing, since that would be a power grab for whatever is trying to do that.
Whatever component bound a handler to an event should be the one unbinding
that handler. But assuming its a good design decision, we could extend
Crafty() to search for entities with a handler for a given event:

Crafty("handles:EventName")

You can then call .each() on the result and unbind the event from the
entities directly.

**Comment by sbyrnes321** [comment]
I agree with @mpetrovic 's proposal. I have no opinion about whether to implement Crafty("handles:EventName") ... I don't _personally_ need that feature, but I don't see anything wrong with it.

**Comment by cawoodm** [comment]
RE: the "power grab" - I find it hard to imagine a scenario in which a developer would globally and willingly unbind all listeners to an event and then regret it because they unbound something they didn't want unbound (and then not know what to do and blame CraftyJS). We shouldn't be protecting ourselves from ourselves in that way. I also find it impractical to keep track of listener functions just so I can later unbind them.

Consider these common scenarios:
1. An object A (JavaScript "Class" instance) is created and listens for an event with `Crafty.bind('Event1', function() {});` Result: absolutely no way to unbind as things stand.
2. An minor object B binds with `Crafty.bind('EventA', this.fcn);` Later object B is dereferenced (in the main code). Result:  the handler and listener still remains so Object B is never destroyed and GC'ed

Not leaving the developer the option to clean the slate with `Crafty.unbind('Event'1)` is taking a powerful tool away from them in the fear that they may do something bad with it. By that reasoning CraftyJS would be reduced to `{};`. Whoever does `Crafty.unbind('Event1')`, does so willingly.

**Comment by mpetrovic** [comment]
And I find it hard to believe someone would have a good reason to bind a
global event handler that could cause a memory leak. Something which goes
against the Entity-Component paradigm. I don't think we even use any of
those in the engine itself. Which makes we wonder why we even have it in
the first place.

Can anyone think of a valid use case that can be better served by binding
to an entity?

**Comment by cawoodm** [comment]
Well the E-C paradigm is notoriously vague and different interpretations are out there... However, if I understand you correctly you suggest that components in Crafty should only be able to bind to events on their own entities or those entities they select by component?

So instead of component X saying `Crafty.bind('EventA')` it should _always_ do either `this.bind('EventA')` or `Crafty('ComponentA').each(function(){this.bind('EventA');})`?

I guess we need to ask whoever wrote `Crafty.bind` but in a lazy way it's nice to just throw an event out to the world and have anyone interested listen to it.

**Comment by mpetrovic** [comment]
that's what Crafty.trigger() does. It throws out an event, and anyone
listening on any entity can catch it. I have no argument with that. Binding
on the global namespace is the issue here.

**Comment by sbyrnes321** [comment]
@cawoodm -- If I understand correctly, your [two scenarios](https://github.com/craftyjs/Crafty/issues/417#issuecomment-13619994) would be satisfactorily solved by [@mpetrovic 's proposal](https://github.com/craftyjs/Crafty/issues/417#issuecomment-13556902), wherein `Crafty.unbind('Event1')` would unbind all the global event listeners for Event1, but not the entity-bound event listeners. Is that correct? If not, could you be more specific?

@mpetrovic -- If I understand correctly, you're asking "What is a use-case where global event binding makes sense?" Lets say I have an event `'GameOver'` that is triggered when the character loses all his life. What kinds of things might this event do? Well, it could trigger the enemy entities to do a victory dance. For that, we would use entity event binding. But there are other things that I might want to happen at GameOver that can't easily be entity-bound. Maybe I want to destroy everything and launch a new scene. Maybe I want to send a message to my server. Maybe I want to play a sound, or make the background flash different colors. It seems to me that global event binding is a convenient way to organize this kind of stuff (assuming there are multiple, independent responses that do not need to occur in any particular order). Obviously there are always other methods to accomplish the same thing. But it seems to me that there are situations where global event bindings are more convenient and productive than any alternative.

**Comment by cawoodm** [comment]
@sbyrnes321, in short I'm happy if we get `Crafty.unbind(eventName)` - my original proposal.

What we're discussing is global binding. I just don't see why componentA has to find the entities it wants to bind to instead of just saying Crafty.bind(). Example:

An entity/component representing the score of a player listens for the events 'Born' and 'Died' using `Crafty.bind()`. Each time a player's unit dies (and there could be thousands of such units) it decrements the score and each time on is born it increments. If I understand mpetrovic correctly I have to constantly loop through thousands of entities and bind to their events. Each time new entities are created I have to update my entity bindings instead of one time running `Crafty.bind()`.

**Comment by sbyrnes321** [comment]
@cawoodm -- your original proposal was for `Crafty.unbind(eventName)` to unbind both global-bound and entity-bound callbacks, and that's what [your pull request](https://github.com/craftyjs/Crafty/pull/418) would do. I was talking about unbinding only global-bound listeners, which was [proposed by Matt](https://github.com/craftyjs/Crafty/issues/417#issuecomment-13556902), and which is implemented by [my pull request](https://github.com/craftyjs/Crafty/issues/420).

So I'll ask more specifically: If [my pull request](https://github.com/craftyjs/Crafty/issues/420) were accepted, would you be totally happy with everything? If not, how would you change it?

**Comment by mpetrovic** [comment]
Because ComponentA should not be binding handlers to other entities' events.

"An entity/component representing the score of a player listens for the
events 'Born' and 'Died' using Crafty.bind()."

Incorrect. The component representing an enemy binds handlers to its own
Born and Died. These handlers get the entity with the Score component (
Crafty('ScoreManager')), and run a method telling it to update itself. The
entire behavior is now added to an entity through a component.

Events are primarily for communication between components of an entity, or
for entities receiving environment level events, such as EnterFrame or
Clicked. That is how they work in Crafty. The global bind, I'm not sure why
it was added originally, but I believe it's best used by non-entities.
Server communication, a physics engine, etc.

**Comment by cawoodm** [comment]
@mpetrovic I've done some tests and `Crafty.trigger` combined with `this.bind` is what I need for my purposes. I want any entity to publish events and any entity (in my case a single Scoreboard or SoundManager) to listen.

I also concluded that `Crafty.bind` is only useful for non-entities. A useful feature which should remain IMO.

@sbyrnes321 I trust your pull request does what you say. It should unbind all listeners originating with `Crafty.bind`. I would find it powerfully useful to also unbind entity-local listeners but not necessary.

**Comment by starwed** [comment]
I'm leaning towards merging @sbyrnes321 patch.  Any last thoughts before I do so?

**Comment by starwed** [comment]
I merged #420, so closing this issue for now.
